### ISSUE TITLE
2.0 Discuss support for clojure.spec

### ISSUE THREAD

**Issue by snoe** [seancorfield/honeysql#146]
I'm raising an issue to discuss the possibility of adding clojure.spec to the project; with a possible goal of a full specification for the input map to `honeysql.core/format`.

I haven't used spec myself yet but as I look at I see benefits to honeysql users and I see an implementation dividing into two parts (which I think is possible). I don't know if it should be part of this project itself or a separate project with support. 

The first part is a sql specification provided by this library covering the format of statements and built-in functions. For example the `:where` value needs to be an expression.

The second is a spec added by end users that could ensure their queries are using valid sql objects (table, column and user-function names). I suspect the library could aid this effort by creating helpers where a user can provide a list of their objects and then augment the built-in specs to return one that's more specific. For example instead of `:select` allowing any column name only allow columns that exist in the `:from` and `:join` tables.

Both parts could be opt-in but I think it would provide a powerful level of safety generally missing from sql generation. 

I also think that most queries people write would fall under a fairly simple specification, say just `:select`, `:from`, `:where` without functions or subselects. A first step could be to add specifications for this most common use case and add support for more complicated queries with later releases.

What does the community think?

**Comment by MichaelBlume** [comment]
I think a separate project, or at least a separate namespace, would be ideal, since clojure.spec hasn't even landed in an official release yet, and even once it does we might not want to immediately make honeysql dependent on it.

**Comment by jrdoane** [comment]
@MichaelBlume, how about just for the ```:dev``` profile in the project.clj for starters? I think there is a benefit to be had when defining specs for dev and testing without impacting the rest of the library. That lets you get started with it but, doesn't force the entire library to depend on it.

For testing purposes, spec'ing out that input map would be an excellent idea. It also opens the possibility of running generative tests which is far more likely to catch bugs that no one had thought about or encountered.

**Comment by cogwheel** [comment]
Adding some support for this. We're using spec throughout our entire stack (Clojure backend, ClojureScript app), especially at the transit level. We've got instrumentation checking arg & ret specs at pretty much every level using some custom defn-like macros to minimize boilerplate. The performance impact has been minimal and the instrumentation can be disabled on an as-needed basis (we're planning to use it in production until it proves to be an issue).

One of the areas of our code that doesn't have any spec coverage right now are the shared queries, custom helpers, etc. We can get away with using `map?` for now, or just use un-checked functions, but ideally there would be specs provided by (alongside) HoneySQL that we could validate against.

**Comment by MichaelBlume** [comment]
I don't have the bandwidth to drive this, but I'd be happy to take pull requests. If you plan to work on this and want to run an approach by me before you get started, I'd also be happy to look at that.

**Comment by seancorfield** [comment]
As a heavy user of spec, I can see value in adding `honeysql.specs` (for example) as an optional namespace that users can require to get access to specs for parts of HoneySQL. I think a spec for the SQL data structure would be valuable, especially for folks who are building data structures directly.

I think the goal of users providing specs for tables and having those integrated with HoneySQL's specs is not achievable yet -- spec doesn't lend itself to programmatic specs yet. Enhancements coming to spec might make that a more tractable goal.

I think creating specs for the data structure is quite a bit of work -- do any of the folks in this thread have the bandwidth to contribute?

**Comment by seancorfield** [comment]
I'm going to assume no one else has the bandwidth to communicate, but I still consider this a worthwhile enhancement but it will be a while before I can get to it.

**Comment by seancorfield** [comment]
I'm developing HoneySQL 2.x at this point and planning to have some level of (optional) Spec support in it. I'm considering writing Specs for every "built-in" piece of syntax but it's made somewhat more difficult by the lack of documentation around the current data DSL so I pivoted from writing specs to writing a new formatting engine (that produces near-identical output for the existing formatting tests) and then I'll go back to writing specs once I believe I understand what the data DSL is meant to look like.

I am undecided whether these Specs will be optional (and thus continuing to support Clojure 1.8 and earlier) or whether I will make Clojure 1.9 a requirement for HoneySQL 2.x (and thus be able to leverage the Specs directly in the new code).

**Comment by seancorfield** [comment]
Per #275 Clojure 1.9 will be the minimum version supported and I plan to document/spec the core data structure supported by v2.

**Comment by seancorfield** [comment]
Having reached Beta 1 with the new formatting engine, it's clear to me that specifying the possible structure of the DSL using Clojure's Spec is pretty much impossible. SQL is sufficiently complex -- and the DSL also sufficiently complex -- that a "complete" Spec would be very complex and likely to produce very poor "explain" messages (and also likely be very slow). A minimal Spec probably could be written but I don't think it would provide much support in use -- most clauses in the DSL accept a sequence of "stuff" (although some clauses support both a sequence and a single entity name for the convenience of use) and that "stuff" can contain be a fairly arbitrary set of things.

TL;DR: Clause: a map from ident to "stuff". Expression: an ident or a value or a sequence. I think the new documentation provides more support than that...

**Comment by seancorfield** [comment]
I should also have mentioned that it's also pretty much impossible to try to restrict the DSL to something that produces valid SQL, given the somewhat ad hoc differences between SQL dialects (just flipping back and forth between SQL Server and PostgreSQL documentation for the same statement was enough to make my head spin -- let alone trying to figure out what is also valid for MySQL and for the many "small" and embedded databases).

**Comment by vemv** [comment]
Hi,

I wonder if it would be reasonable to reopen this issue under a humbler goalpost?

It is easy to comprehend that SQL can't be reasonably speced as a whole. However that doesn't remove the possibility of specing various defns of honeysql's impl.

To give a concrete example, today I faced this stacktrace:

```
   Don't know how to create ISeq from: clojure.lang.Keyword

                   RT.java:  557  clojure.lang.RT/seqFrom
                   RT.java:  537  clojure.lang.RT/seq
                  core.clj:  139  clojure.core/seq
                  core.clj: 7294  clojure.core/partition-all/fn
              LazySeq.java:   42  clojure.lang.LazySeq/sval
              LazySeq.java:   51  clojure.lang.LazySeq/seq
                   RT.java:  535  clojure.lang.RT/seq
                  core.clj:  137  clojure.core/seq
             protocols.clj:   24  clojure.core.protocols/seq-reduce
             protocols.clj:   75  clojure.core.protocols/fn
             protocols.clj:   75  clojure.core.protocols/fn
             protocols.clj:   13  clojure.core.protocols/fn/G
                  core.clj: 6853  clojure.core/reduce
                  core.clj: 6835  clojure.core/reduce
                  sql.cljc:  444  honey.sql$format_join/invokeStatic
```

Nothing too bad, one can quickly make sense of it (expected a coll, got keyword). Perhaps beginners not so much though.

So what I'd propose is that `honey.sql/format-join` had a discrete `fdef`, making a simple specification of the shape of its inputs.

From intuition, I think that a reasonable number of honey impl defns could get such a `fdef` without falling into rabbit holes.

If things got too recursive etc one could choose to use coarse-grained specs such as `coll?`. One still gets a better error message than bare clojure.core stacktrace, which rarely tell you _the value at fault_.

Relatedly, the fdefs could grow from sparse and unsophisticated to something more comprehensive. So that one doesn't need a large upfront investment.

WDYT?

Cheers - V

**Comment by seancorfield** [comment]
I'd be open to PRs to add a new `honey.sql.specs` ns with `instrument`/`unstrument` fns (see `next.jdbc` for a reference model), that provides optional Specs for some internal parts of HoneySQL.

Since HoneySQL requires a minimum Clojure 1.9, I'd be OK with `honey.sql` requiring that `specs` ns and including the `fdef`'s inline above each definition (so it is reasonable/easy to Spec private functions).

Any such PRs should also update the tests to require `honey.sql.specs` and call `instrument` (again, as `next.jdbc`'s tests do).

**Comment by vemv** [comment]
Nice, will keep in mind!

Probably for a couple months I'll see if I can implement a few fdefs that would help me in my daily work, defining them outside of honeysql for easy iteration. I'd contribute them as the last step.

If someone else beats me to contributing a PR I'll be happy as well.
